### ISSUE TITLE
why 253 can not beat 249

### ISSUE THREAD

**Issue by hilbertcao** [leela-zero/leela-zero#2541]

**Comment by Friday9i** [comment]
Hypothesis: because the selection process is extremely doubtful?
Why is it potentially doubtful? Because almost all match games have exactly the same long opening, leading to a huge bias: candidate nets are evaluated only against current best net and only after a single long opening, ie not at all in a "broad panel of representative go situations" (which is what we want of course)... Hence, a promoted net can convincingly beat its predecessor in a match without being any better (or even worse) at Go, in a broad panel of pertinent go situations...
This can directly lead to this bad result: LZ253 is not any better than LZ249 (despite beating LZ252 in match conditions, itself beating LZ251, ...).
And we know that for exactly 1 year... Eg https://github.com/leela-zero/leela-zero/issues/2047
Unfortunately, nothing has changed since that. I proposed alternatives, such as https://github.com/leela-zero/leela-zero/issues/2104, or varying visits in match games, or introducing some random noises in them, or doing like Deepmind did for AlphaChess (using a pool of previous nets, and/or chosing in openings alternate moves within 1% winrate of preferred move), etc. Unfortunately, nothing has changed since that: LZ still use this (probably) biased process for promotions... 
Maybe this bad result will finally move things a bit, who knows?

**Comment by Splee99** [comment]
In my own test matches I did see some evolution in playing style. When LZ improves its solid defense technique, it doesn't mean its overall winrate increases. However, I did see more counter attacks (instead of solid defense) were played in recent promotions. For higher visits, the winrate against ELF is now higher compared to previous LZ versions in same game setup.

**Comment by Friday9i** [comment]
I'm not saying it it not progressing overall (I demonstrated months ago an average improvement of ~5 Elo per new net, cf https://github.com/leela-zero/leela-zero/issues/2425. It's probably quite true today, if someone is willing to test).
I'm just saying that progression is very slow and quite chaotic, probably in part due to this bias in selection process. Pity to spend significant time to tests candidate nets, based on a biased process (bias means the result cannot really be trusted).
And there is worse: this biased process leads to the rejection ~90% of candidate nets, while many of them must be quite good... More than time to rethink the process and try alternatives IMHO

**Comment by l1t1** [comment]
newer lz nets winrate vs elfv2 v800 is dropping
248 was 0.66, 253 is only 0.5
see http://www.yss-aya.com/cgos/19x19/standings.html

**Comment by tapsika** [comment]
>  the rejection ~90% of candidate nets, while many of them must be quite good

You said this in the past, but why? I see nothing pointing in this direction. Maybe you can demonstrate by finding a few such good nets? Otherwise we only know (from reruns) that promoted nets are only marginally better, most promoted on luck only. So no real, "quite good" promotions like in the past (with significant strength gains) - thus presumably neither among failed candidates. And therefore nothing in 47%-53% is surprising in a single (re)match. But OC after a few steps 47% is not nice.

**Comment by Friday9i** [comment]
Yes and no...: a net getting a real 52% in a proper match with proper diversity and an infinite number of games will get anywhere between ~47% and 57% win rate in a non-biased 400-games match, and a larger spread of winrate if the match is biased (ie LZ's match conditions)... Hence, a "true" 52% net has probably 30% to 40% chance of being rejected... Hence, we must reject many better candidates, which explains the sentence "rejection ~90% of candidate nets, while many of them must be quite good"

**Edit:** a "true" 52% net has ~70% chances of being rejected as it has to be lucky and get 55% to be selected. And because it seems that candidate with true winrate >55% must be extremely rare (otherwise we would get some 60+% selected nets from time to time, but we never had it!), it means that we reject very probably significantly more than 50% of the promising candidates ;-(

**Comment by tapsika** [comment]
This is not specific to current LZ's state: the condition of "55% after 400 games" is never meant to catch ALL >50% nets, but to MOSTLY catch >50% nets.

For reliably testing a single (strengthwise close) net pair 400 games is obviously nowhere enough, since at the very least you need sd <1% or 0.5% for confidence (which doesn't worth it).

But my doubts about your claim are for the "quite good" part - I think most candidates are slightly >50% in reality.

**Comment by Friday9i** [comment]
I guess real candidates are between ~45% and ~52%, hence I consider that a candidate >50% is "quite good" ;-). Hence, many candidates (~30% of nets?) should be "quite good" but are rejected.
Anyway, "things should be tested", that's my main point... Eg 50% selection threshold (it works well for KatagGo and is supported by an experiment I did here https://github.com/leela-zero/leela-zero/issues/1524, showing it is probably more efficient:  done 1.5 years ago), no-gating (probably less efficient, but why not try it), introducing noise in match games, ...
Note: the gating simulation exactly takes into account what you say : -)
Whatever, but things should be tested IMHO at this point, rather than letting LZ slowly dying ; -(

**Comment by diadorak** [comment]
https://www.cloudygo.com/leela-zero-eval/eval-graphs 

LZ253 is doing very well.

**Comment by Friday9i** [comment]
Indeed, seems LZ253 is quite good, and this confirms that LZ is still improving ~4 or 5 Elo per new net on average, so some improvement (but way below the ~35 Elo boost per net from the ~55% winrate.  Not a surprise to be below, but 7x or 8x below is a lot).
Relaxing the selection process could help, that's very easy to test, more nets will have a chance to tests new approaches and the risk of regression by selecting nets with eg 53% winrate instead of 55% is very low... My theoretic tests were even showing than optimal threshold is probably between 50% to 52%, but they cannot take into account the retroaction of the selection process on the strength of candidate nets (that has to be tested, no alternative), so we should be prudent and test preferably 52% or 53% rather than 48% or 50% (unless we are risk-takers : -).
Anyway, I proposed it 10 times without success, I doubt anything will be tested: "no hard proof, we don't change anything" (even though multiple experiences were done successfully with 50% gating or, less succesfully, without gating)

**Comment by l1t1** [comment]
254 seems better on cgos
http://www.yss-aya.com/cgos/19x19/standings.html

**Comment by Marcin1960** [comment]
@Friday9i " I proposed it 10 times without success, I doubt anything will be tested: "no hard proof, we don't change anything" 

I also would like to see the 50% gating and forcing some variations during the openings. It would open new ways for the net. Last year provided hard proof that present approach does not work.

**Comment by Vargooo** [comment]
> Indeed, seems LZ253 is quite good, and this confirms that LZ is still improving ~4 or 5 Elo per new net on average, 

It seems right.
I've just run a 100 game match LZ254 v LZ204  (-v 1601 --noponder)
LZ 254 wins 82-18 , so 82% (if someone wants the games, I can upload them)

According to https://wismuth.com/elo/calculator.html#elo_diff=250&system=goratings
a 250 Elo difference (5Elo x 50 networks) gives around 81% winrate
So,  around 250 "real" Elo gained since February

**Comment by Marcin1960** [comment]
@Vargooo around 250 "real" Elo gained since February

Do you think that this gain will be repeated over the next year or two?

**Comment by Splee99** [comment]
I'm still thinking we should increase the visits in self play games, which means updating the autogtp to generate higher quality games slowly. I hope in this way the overall learning speed will be faster. Just imagine a learning process of 1 visit only, how long would it take to train LZ 40b from scratch? Probably much much longer than the current one.

**Comment by l1t1** [comment]
sai games visits 2400 now

**Comment by nemja** [comment]
More visits improves the quality somewhat, but helps little with blind spots that are not explored except maybe at extremely high visits. And selfplay level will never be higher than a certain limit because of misplayed ladders and such, which continue to teach the net wrong habits and incorrect moves so it won't learn everything it could theoretically. More visits helps with some ladders, but are useless with others (where the net wrongly thinks it doesn't work so search doesn't explore it).

What LZ lacks regarding selfplay visit optimum is experimentation.

**Comment by SHKD13** [comment]
@Vargooo 
I'd like to take a look at these sgf's of LZ 254 vs 204

**Comment by Vargooo** [comment]
https://drive.google.com/open?id=1dUrKsrMprFg3Bc8Ou3UTfHyJrwftFzoN
https://drive.google.com/open?id=1lzmFWyVUcEzvvE4OUTHSE4J4ykDhXqew
First link : 70 games, second link : 30 games

**Comment by Crazyeight101** [comment]
It seems a lot of wheel spinning has been going on since I was here last.

@friday9i Has anyone created a heat map of the relative weights between networks that we can look at while comparing match results and elos?  Having a clear visual representation of the networks will go a long way towards making a case for any changes to the pipeline.  For instance, we'd be able to confirm the existence or nonexistence of cyclical patterns over time or biases in training data in a detailed but easily understood way.
